xbuild/src/cargo/mod.rs

```diff
@@ -101,7 +101,7 @@ impl Cargo {
         artifact: Option<Artifact>,
         ty: CrateType,
     ) -> Result<PathBuf> {
-        let arch_dir = if target.platform() == Platform::host()? && target.arch() == Arch::host()? {
+        let arch_dir = if target.is_host()? {
             target_dir.to_path_buf()
         } else {
             target_dir.join(target.rust_triple()?)
```

Comment:
the reason for that is we set env variables for different targets, causing cargo to think the build is out of date when compiling for a different target. this way cargo caches the builds properly

Reply:
We've also had slowdowns with this with `cargo-apk` when building Android and host-native binaries in the same `target/` even though everything non-host ends up in `target/<triple>` and shouldn't touch `target/`, you're saying that the env vars are the reason for these seemingly unnecessary rebuilds?

Reply:
yes, cargo stores the env vars as inputs to the build globally, not per target. if it were stored per target as long as you always use `x`, we wouldn't have to override the `target-dir`. but I'd still put it in `target/x/target`, to avoid accidentally running `cargo build` and then having to rebuild from scratch wenn running `x build`.